go-controller/pkg/ovn/ovn.go

```diff
@@ -694,39 +694,40 @@ func (oc *Controller) GetNetworkPolicyACLLogging(ns string) *ACLLoggingLevels {
 	return &nsInfo.aclLogging
 }
 
-// Verify if controller can support ACL logging and validate annotation
-func (oc *Controller) aclLoggingCanEnable(annotation string, nsInfo *namespaceInfo) bool {
-	if !oc.aclLoggingEnabled || annotation == "" {
-		nsInfo.aclLogging.Deny = ""
-		nsInfo.aclLogging.Allow = ""
-		return false
-	}
+// aclLoggingUpdateNsInfo parses the provided annotation values and sets nsInfo.aclLogging.Deny and
+// nsInfo.aclLogging.Allow.
+// Parse the provided annotation string. Valid values for "allow" and "deny" are  "alert", "warning", "notice", "info",
+// "debug" and "".
+// In the following special cases, nsInfo.aclLogging.Deny and nsInfo.aclLogging.Allow. will both be reset to "", meaning
+// that logging will be switched off:
+// i) oc.aclLoggingEnabled == false
+// ii) annotation == ""
+// iii) annotation == "{}"
+// Also, if one of "allow" or "deny" can be parsed, but the other key is not present in the annotation, then assume
+// that this key should be disabled by setting its nsInfo value to "".
+// Return an error if the provided annotation cannot be unmarshaled.
+func (oc *Controller) aclLoggingUpdateNsInfo(annotation string, nsInfo *namespaceInfo) error {
 	var aclLevels ACLLoggingLevels
-	err := json.Unmarshal([]byte(annotation), &aclLevels)
-	if err != nil {
-		return false
+	// If logging is disabled or if the annotation is "" or "{}", use empty strings. Otherwise, parse the annotation.
+	if oc.aclLoggingEnabled && annotation != "" && annotation != "{}" {
+		err := json.Unmarshal([]byte(annotation), &aclLevels)
+		if err != nil {
+			return fmt.Errorf("could not unmarshal namespace ACL annotation '%s', err: %q", annotation, err)
+		}
 	}
 
-	// Using newDenyLoggingLevel and newAllowLoggingLevel allows resetting nsinfo state.
-	// This is important if a user sets either the allow level or the deny level flag to an
-	// invalid value or after they remove either the allow or the deny annotation.
-	// If either of the 2 (allow or deny logging level) is set with a valid level, return true.
-	newDenyLoggingLevel := ""
-	newAllowLoggingLevel := ""
-	okCnt := 0
-	for _, s := range []string{"alert", "warning", "notice", "info", "debug"} {
-		if s == aclLevels.Deny {
-			newDenyLoggingLevel = aclLevels.Deny
-			okCnt++
+	validLogLevels := sets.NewString("alert", "warning", "notice", "info", "debug", "")
+	if validLogLevels.Has(aclLevels.Deny) {
+		if nsInfo.aclLogging.Deny != aclLevels.Deny {
```

Comment:
maybe it will be easier to read if you just do the assignment without additional comparison?

Reply:
Yeah, I always do that, I don't know why :-/